### PR TITLE
[clang][dataflow] When analyzing ctors, don't initialize fields of `*this` with values.

### PR DESCRIPTION
**Pull request by martinboehme**
This is the constructor's job, and we want to be able to test that it does this.

### CODE REVIEW

**Review by ymand — APPROVED** [review]

**Review by Xazax-hun — APPROVED** [review]

File: clang/lib/Analysis/FlowSensitive/DataflowEnvironment.cpp [hunk 1]
```diff
@@ -799,6 +806,16 @@ PointerValue &Environment::getOrCreateNullPointerValue(QualType PointeeType) {
   return DACtx->getOrCreateNullPointerValue(PointeeType);
 }
 
+void Environment::initializeFieldsWithValues(RecordStorageLocation &Loc) {
```
[the remark this answers is not in the document]
Comment (martinboehme) [reply clang/lib/Analysis/FlowSensitive/DataflowEnvironment.cpp:809]: I consciously put this here to be consistent with the ordering in the header file. (Do we have a firm convention for this?)

File: clang/lib/Analysis/FlowSensitive/DataflowEnvironment.cpp [hunk 2]
```diff
@@ -414,8 +414,15 @@ void Environment::initialize() {
...
     } else if (MethodDecl->isImplicitObjectMemberFunction()) {
       QualType ThisPointeeType = MethodDecl->getFunctionObjectParameterType();
-      setThisPointeeStorageLocation(
-          cast<RecordStorageLocation>(createObject(ThisPointeeType)));
+      auto &ThisLoc =
+          cast<RecordStorageLocation>(createStorageLocation(ThisPointeeType));
+      setThisPointeeStorageLocation(ThisLoc);
+      refreshRecordValue(ThisLoc, *this);
+      // Initialize fields of `*this` with values, but only if we're not
+      // analyzing a constructor; after all, it's the constructor's job to do
```
[the remark this answers is not in the document]
Comment (martinboehme) [reply clang/lib/Analysis/FlowSensitive/DataflowEnvironment.cpp:422]: Do you mean something like this?

```cxx
struct S {
    S() {}
    int i = 1;
};
```

If so, the answer is yes:

```
  |-CXXConstructorDecl <line:2:5, col:10> col:5 S 'void ()' implicit-inline
  | |-CXXCtorInitializer Field 0xc1eeb90 'i' 'int'
  | | `-CXXDefaultInitExpr <col:5> 'int' has rewritten init
  | |   `-IntegerLiteral <line:3:13> 'int' 1
  | `-CompoundStmt <line:2:9, col:10>
```

https://godbolt.org/z/oYP66fMjT

[the remark this answers is not in the document]
Comment (ymand) [reply clang/lib/Analysis/FlowSensitive/DataflowEnvironment.cpp:422]: Yes. thanks

File: clang/lib/Analysis/FlowSensitive/DataflowEnvironment.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (ymand) [reply clang/lib/Analysis/FlowSensitive/DataflowEnvironment.cpp:809]: I don't know of any convention. I tend not to anchor on header order, but YMMV.  Up to you.

**Review by ymand — APPROVED** [review]

**Pull request merged**